daily_read/order_portal.py

```diff
@@ -76,20 +89,53 @@ def process_orders(self, closed_before_in_days=30):
             if order["identifier"] in self.projects_data.data.keys():
                 proj_info = self.projects_data.data[order["identifier"]]
                 if order["reports"]:
-                    proj_prog_rep = next(item for item in order["reports"] if item["name"] == "Project Progress")
-                    proj_info.report_iuid = proj_prog_rep["iuid"]
-                if proj_info.orderer in order_updates:
-                    order_updates[proj_info.orderer]["projects"].append(proj_info)
-                else:
+                    prog_reports = [item for item in order["reports"] if item["name"] == "Project Progress"]
+                    if len(prog_reports) > 1:
+                        log.error(f"Multiple reports for Project Progress for order {order['identifier']}")
+                    else:
+                        proj_info.report_iuid = prog_reports[0]["iuid"]
+
+                if proj_info.orderer not in order_updates:
                     order_updates[proj_info.orderer] = {
                         "pull_date": f"{pull_date}",
-                        "projects": [proj_info],
+                        "active_projects": 0,
+                        "recents": {},
+                        "projects": {},
                     }
 
+                latest_date = "0000-00-00"
+                latest_status = "None"
+                for date_value in proj_info.data["proj_dates"]:
+                    date_statuses = proj_info.data["proj_dates"][date_value]
+                    # Activity on 5 recent dates
+                    recents_keys = sorted(order_updates[proj_info.orderer]["recents"].keys())
+                    if len(recents_keys) < 5:
+                        for status in date_statuses:
+                            order_updates[proj_info.orderer]["recents"].setdefault(date_value, []).append(
+                                (proj_info.data["project_name"], status)
+                            )
+                    else:
+                        if date_value > recents_keys[0]:
+                            order_updates[proj_info.orderer]["recents"].pop(recents_keys[0])
+                            for status in date_statuses:
+                                order_updates[proj_info.orderer]["recents"].setdefault(date_value, []).append(
+                                    (proj_info.data["project_name"], status)
+                                )
+
+                    # Find status
+                    if date_value > latest_date:
+                        latest_date = date_value
+                        for status in date_statuses:
+                            if dates_prio[status] > dates_prio[latest_status]:
```

Comment:
Why do we need `dates_prio` if we're only looking at it for the statys with the very latest date?

Reply:
From anand:
The data in proj_info.data["proj_dates"]  has dates as keys
So if 2 statuses have the same date, it would be 'date': [status1, status2]